src/main/java/com/amazon/opendistroforelasticsearch/ad/task/ADTaskCacheManager.java

```
@@ -0,0 +1,306 @@
+    }
+
+    /**
+     * Set threshold model trained or not.
+     *
+     * @param taskId task id
+     * @param trained threshold model trained or not
+     */
+    public void setThresholdModelTrained(String taskId, boolean trained) {
+        ADBatchTaskCache taskCache = getBatchTaskCache(taskId);
+        taskCache.setThresholdModelTrained(trained);
+        if (trained) {
+            int size = taskCache.getThresholdModelTrainingData().size();
+            long cacheSize = trainingDataMemorySize(size);
+            taskCache.getThresholdModelTrainingData().clear();
+            taskCache.getCacheMemorySize().getAndAdd(-cacheSize);
+            memoryTracker.releaseMemory(cacheSize, false, HISTORICAL_SINGLE_ENTITY_DETECTOR);
+        }
+    }
+
+    /**
+     * Get shingle data.
+     *
+     * @param taskId AD task id
+     * @return shingle data
+     */
+    public Deque<Map.Entry<Long, Optional<double[]>>> getShingle(String taskId) {
+        return getBatchTaskCache(taskId).getShingle();
+    }
+
+    /**
+     * Check if task exists in cache.
+     *
+     * @param taskId task id
+     * @return true if task exists in cache; otherwise, return false.
+     */
+    public boolean contains(String taskId) {
+        return taskCaches.containsKey(taskId);
+    }
+
+    /**
+     * Check if there is task in cache for detector.
+     *
+     * @param detectorId detector id
+     * @return true if there is task in cache; otherwise return false
+     */
+    public boolean containsTaskOfDetector(String detectorId) {
+        return taskCaches.values().stream().filter(v -> Objects.equals(detectorId, v.getDetectorId())).findAny().isPresent();
+    }
+
+    /**
+     * Get batch task cache. If task doesn't exist in cache, will throw
+     * {@link java.lang.IllegalArgumentException}
+     *
+     * @param taskId task id
+     * @return AD batch task cache
+     */
+    private ADBatchTaskCache getBatchTaskCache(String taskId) {
```

runtime exception means sth you cannot recover.  Is getting something not existing not recoverable? Is it better to return an optional.empty when it is not there.

---

This is to make the error handling easier in `TaskManager` (I will send out another PR soon). In task manager, we will catch all exceptions, differentiate customer input error and server error in one place. We definitely can return `opotional.empty` or `null` to tell `TaskManger` that this task is not found in cache. But we need extra logic to handle such not found exceptions and in task runner, we need to check task existence for every pieces then remember to return it when the task not found. So I prefer to throw exception here and catch exceptions in one place, rather than check the returned result and check if it's `Opitonal.empty` or `null` in multiple places.

---

Can you explain these in comments?  In a similar place where Weicong commented, it is uncommon to write get/put by throwing exceptions in this way.  Others will want to know this when they read/modify these methods.

---

Oh, I didn't see this comment either as it's resolved before I addressed this. Will add some comments in java doc.